### PR TITLE
Support composing with input source maps

### PR DESCRIPTION
**Pull request by crisptrutski**
Adds option `.inputSourceMap`; when set to a SourceMap object, the output sourcemap is composed with the inputSourceMap.

### CODE REVIEW

**Comment by crisptrutski** [comment]
@johnjbarton added test, and tweaked some edge cases. 

Guessing you'll also want a documented CLI option, but new IO to load the source map file, and until `addOptions` moves from `src/Options.js` to `src/node.js` that seems problematic. Skip for now?

File: src/Compiler.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ import {
 
 import {ParseTreeMapWriter} from './outputgeneration/ParseTreeMapWriter.js';
 import {ParseTreeWriter} from './outputgeneration/ParseTreeWriter.js';
+import {SourceMapConsumer} from './outputgeneration/SourceMapIntegration.js';
```
Comment (arv) [inline src/Compiler.js]:
```js
import {
  SourceMapConsumer,
  SourceMapGenerator
} from './outputgeneration/SourceMapIntegration.js';
```

File: test/unit/codegeneration/SourceMap.js [hunk 2]
```diff
@@ -91,6 +90,65 @@ suite('SourceMap.js', function() {
...
   });
 
+  test('SourceMap with input map', function() {
+    var src = 'function foo() { return 5; }\nvar \nf\n=\n5\n;\n';
+    var srcLines = src.split('\n');
+    var filename = 'sourceMapThis.js';
+
+    // maps every character one column right and two lines down
+    var g = new SourceMapGenerator({file: 'sourceMapThis.js'});
+    for (var i=0; i<srcLines.length; i++) {
```
Comment (arv) [inline test/unit/codegeneration/SourceMap.js]: ws after = and <

File: test/unit/codegeneration/SourceMap.js [hunk 3]
```diff
@@ -91,6 +90,65 @@ suite('SourceMap.js', function() {
...
+        g.addMapping({
+          generated: {line: i + 1, column: j},
+          original: {line: i + 3, column: j + 1},
+          source: 'priorSource.js'
+        });
+      }
+    }
+    var inputMap = g.toJSON();
+
+    var compiler = new Compiler({sourceMaps: 'file',
```
Comment (arv) [inline test/unit/codegeneration/SourceMap.js]: Indent like this

```js
 var compiler = new Compiler({
  sourceMaps: 'file',
  script: true,
  inputSourceMap: inputMap
});
```

File: test/unit/codegeneration/SourceMap.js [hunk 4]
```diff
@@ -91,6 +90,65 @@ suite('SourceMap.js', function() {
...
+      // function foo() { return 5; >}<
+      {generated: {line: 3, column: 0}, original: {line: 3, column: 28}},
+      {generated: {line: 4, column: 0}, original: {line: 5, column: 1}},
+      {generated: {line: 5, column: 1}, original: {line: 8, column: 1}}
+    ];
+
+    testcases.forEach(function(testcase, caseNumber) {
+      var actual = consumer.originalPositionFor(testcase.generated);
+      var shouldBeTrue = actual.line === testcase.original.line;
+      assert.isTrue(shouldBeTrue, caseNumber + ' Line mismatch ' + actual.line);
```
Comment (arv) [inline test/unit/codegeneration/SourceMap.js]: assert.equal or assert.strictEqual and you can skip the _message_.

**Comment by arv** [comment]
LGTM with nits

**Comment by crisptrutski** [comment]
@arv thanks for the feedback. Have also rebased on master for clean merge

**Comment by guybedford** [comment]
:+1: this would be amazing for the next SystemJS release, allowing JSX -> ES6 plugins with source maps.

File: test/unit/codegeneration/SourceMap.js [hunk 5]
```diff
@@ -83,12 +82,8 @@ suite('SourceMap.js', function() {
 
     testcases.forEach(function(testcase, caseNumber) {
       var actual = consumer.originalPositionFor(testcase.generated);
-      var shouldBeTrue = actual.line === testcase.original.line;
-      assert.isTrue(shouldBeTrue, caseNumber + ' Line mismatch ' + actual.line);
-      var expected = testcase.original.column;
-      shouldBeTrue = actual.column === expected;
-      assert.isTrue(shouldBeTrue,
-          caseNumber + ' Column mismatch ' + actual.column + ' vs ' + expected);
+      assert.strictEqual(actual.line, testcase.original.line);
```
Comment (arv) [inline test/unit/codegeneration/SourceMap.js]: FYI, we usually use `assert.equal` unless we really need strict since it reads better. (I do believe chai made the wrong choice here though.)

**Comment by arv** [comment]
LGTM, landing
